### PR TITLE
Avoid declaring a fake dependency edge

### LINKED ISSUE #68149: rustbuild: Both stage0 and stage1 std are rebuilt spuriously
I'm not sure when exactly it happens, but it reproduces roughly like this:

- Rebase my working branch on upstream/master.
- Run `./x.py test --stage 1 src/test/ui --test-args mytest`, the build starts from `stage0 std` or later and succeeds.
- Run `./x.py test --stage 1 src/test/ui --test-args mytest` again, should be a trivial build, but the build starts from `stage0 std` or `stage1 std` again and succeeds.
- Run `./x.py test --stage 1 src/test/ui --test-args mytest` again, trivial build happens.

The target is x86_64-pc-windows-gnu.
The issue exists for several weeks, so it's a regression.
cc @Mark-Simulacrum

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
When we're producing an rlib, we do not need anything more than an rmeta file
for each of our dependencies (this is indeed utilized by Cargo for pipelining).
Previously, we were still storing the paths of possible rlib/dylib crates, which
meant that they could still plausibly be accessed. With -Zbinary-dep-depinfo,
that meant that Cargo thought that rustc was using both the rlib and an (earlier
emitted) rmeta, and so needed a recompile, as the rlib may have finished writing
*after* compilation started (for more detail, see issue 68149).

This commit changes metadata loading to not store the filepaths of dylib/rlib if
we're going to end up creating an rlib only.

Fixes #68149.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
To be honest I'm not super happy with this change - feels a bit ad-hoc and I worry it could change under us and we wouldn't notice. Not sure we can do better though.

cc @petrochenkov as well

**Comment by ehuss** [comment]
Can you explain more how changing the order the files are printed should fix the issue?

**Comment by Mark-Simulacrum** [comment]
Oh, I was over eager in a refactor, the code should call next() on the iterator instead of pushing the whole iterator on.

**Comment by Mark-Simulacrum** [comment]
Fixed.

**Comment by alexcrichton** [comment]
I think an alternate solution, which may be more of an "at the source" fix, might be to avoid reading these files entirely? For example if we find `*.rmeta` and `*.rlib` for a dependency, and we're producing an rlib, then in theory rustc shouldn't even touch the `*.rlib`. In rustc's own internal data structures we could avoid even listing the path to the `*.rlib`.

I feel like this sort of route would protect against potential bugs where we don't say we depend on the `rlib` but rustc still has access to it and could accidentally use it. I'd prefer if we simply don't give rustc access to the `rlib` internally by not listing the path for the rlib if we can.

**Comment by Mark-Simulacrum** [comment]
I can try to experiment with that path -- I think @petrochenkov planned on trying to do so based on comments on the issue. I do think it's better :)

**Comment by Mark-Simulacrum** [comment]
Initial implementation for the proposed "at the source" change is up, though I've not run tests locally.

File: src/librustc_metadata/locator.rs [hunk 1]
```diff
@@ -681,12 +683,21 @@ impl<'a> CrateLocator<'a> {
...
+        // and this is arguably a bug (see #10786), but because reading metadata
+        // is quite slow (especially from dylibs) we currently do not read it
+        // from the other crate sources.
         if slot.is_some() {
-            // FIXME(#10786): for an optimization, we only read one of the
-            //                libraries' metadata sections. In theory we should
-            //                read both, but reading dylib metadata is quite
-            //                slow.
-            if m.is_empty() {
+            if m.is_empty() || self.sess.crate_types.borrow() == &[config::CrateType::Rlib] {
```
Comment (petrochenkov) [inline src/librustc_metadata/locator.rs]: Hmm, we don't have a separate `CrateType::Rmeta`.
If the build is supposed to produce a `rmeta` file and stop we should act in the same way and return `None` regardless of `crate_types`.
I.e. this logic needs to also consider `sess.opts.output_types` somehow (which can be factored out into `Session::we_only_need_metadata_not_object_code`).

**Comment by petrochenkov** [comment]
This is what I wanted to do, basically.
Except I wanted to write something like `self.sess.crate_types.borrow().all(|crate_type| crate_type == CrateType::Rlib || crate_type == CrateType::Rmeta)`, but it turns out there's no `CrateType::Meta`.

**Comment by petrochenkov** [comment]
I've just realized that if the dependency is a proc macro, then in addition to its `rmeta` we also need its `dylib` early, for `dlopen`ing. That makes it a dependency, in theory.

(Same applies to plugin dependencies, but `-Zbinary-dep-depinfo` doesn't produce depinfo for them at all (which is incorrect, but nobody cares).)

**Comment by Mark-Simulacrum** [comment]
I will refactor/correct the code as you've suggested.

I believe it is indeed the case that we don't care about plugin dependencies (yet, at least). I will not try to solve that problem in this PR.

**Comment by Mark-Simulacrum** [comment]
The failing test case (https://github.com/rust-lang/rust/blob/master/src/test/ui/rmeta-rpass.rs) appears to assert that when we find an rmeta and rlib in the linker path, we should prefer the rlib. I'm not sure why this is true.

Per https://github.com/rust-lang/rust/issues/10786, we would generally want this to be an error (as the metadata in the two files is incompatible), but I don't think we should load both -- I consider it in an unstable implementation detail which file is ultimately used if you have incompatible files.

File: src/librustc_metadata/locator.rs [hunk 2]
```diff
@@ -656,14 +656,34 @@ impl<'a> CrateLocator<'a> {
...
         };
         slot.map(|(svh, metadata)| (svh, Library { source, metadata }))
     }
 
+    fn needs_crate_flavor(&self, flavor: CrateFlavor) -> bool {
+        if flavor == CrateFlavor::Dylib && self.is_proc_macro == Some(true) {
+            return true;
+        }
+
+        let is_rlib = self.sess.crate_types.borrow() == &[config::CrateType::Rlib];
```
Comment (petrochenkov) [inline src/librustc_metadata/locator.rs]:
```suggestion
        let is_rlib = self.sess.crate_types.borrow().all(|crate_type| crate_type == config::CrateType::Rlib);
```
Technically `rustc --crate-type=rlib --crate-type=rlib lib.rs` is legal and I don't see `crate_types` deduplicated anywhere.

File: src/librustc_metadata/locator.rs [hunk 3]
```diff
@@ -681,12 +701,22 @@ impl<'a> CrateLocator<'a> {
...
+        // and this is arguably a bug (see #10786), but because reading metadata
+        // is quite slow (especially from dylibs) we currently do not read it
+        // from the other crate sources.
         if slot.is_some() {
-            // FIXME(#10786): for an optimization, we only read one of the
-            //                libraries' metadata sections. In theory we should
-            //                read both, but reading dylib metadata is quite
-            //                slow.
-            if m.is_empty() {
+            if m.is_empty() && !self.needs_crate_flavor(flavor) {
```
Comment (petrochenkov) [inline src/librustc_metadata/locator.rs]:
```suggestion
            if m.is_empty() || !self.needs_crate_flavor(flavor) {
```
?

**Comment by petrochenkov** [comment]
>The failing test case (https://github.com/rust-lang/rust/blob/master/src/test/ui/rmeta-rpass.rs) appears to assert that when we find an rmeta and rlib in the linker path, we should prefer the rlib. I'm not sure why this is true.

From the PR introducing the test it looks like just something to ensure that the behavior doesn't change in unnoticed way.

**Comment by petrochenkov** [comment]
>I consider it in an unstable implementation detail which file is ultimately used if you have incompatible files.

I wanted to write some short overview of this situation and dependencies, perhaps this weekend.

Right now if two versions of the "same library", e.g. "rmeta vs rlib" or "target vs host" in case of "dual proc macros" become unsynchronized, it's pretty much undefined behavior.
Perhaps we can give some guarantees in this case by loading and comparing fingerprints rather than whole metadata, or perhaps more, need to think a bit more.

**Comment by petrochenkov** [comment]
Before r+'ing I'd like to get a confirmation from Cargo people that something like
```sh
cargo check
modify lib.rs
cargo build
```
won't produce unsynchronized rmeta (older) and rlib (newer) in the same directory.

It looks like the disambiguators (like `2226b7d67c247d50`) in rmeta and rlib file names are different if the source was modified, so the mismatch shouldn't be possible, but I don't know the rules exactly.

**Comment by petrochenkov** [comment]
r=me otherwise

File: src/librustc_metadata/locator.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply src/librustc_metadata/locator.rs]: Crazy! But indeed true.

I've done a bit of an audit of where we use crate_types and a few interesting things turned up:

In save analysis, we seem to write out the output path, but it seems not obviously wrong at least: https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_save_analysis/lib.rs#L90

Interestingly in https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_interface/passes.rs#L371 it seems that `--crate-type proc-macro --crate-type proc-macro` will actually error today.

And in https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_interface/passes.rs#L480 we will generate duplicate output filenames...

https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_codegen_ssa/back/write.rs#L767 will consider `--crate-type rlib --crate-type rlib` to not be an rlib, which means you can get full LTO even in rlibs.

(possibly some more)

Generally speaking it looks like dedup'ing these is a good idea so I've filed https://github.com/rust-lang/rust/pull/68403

**Comment by Mark-Simulacrum** [comment]
Let's r? @alexcrichton then to hopefully resolve the Cargo question (or maybe @ehuss knows better, not sure :)

**Comment by alexcrichton** [comment]
@bors: r=petrochenkov

Code here looks great to me, thanks!

> Before r+'ing I'd like to get a confirmation from Cargo people ...

I believe this is correct, due to all the trickiness involved `cargo check` effectively doesn't share a cache at all with `cargo build`. We try to keep these all separate.

In any case if I'm remembering incorrectly I suspect we'll figure out quickly!

**Pull request merged**